### DISCUSSION TITLE
Upsert returning null value error

### DISCUSSION THREAD

**Discussion by jayehernandez** [supabase/supabase#15730] [Questions] [answered]
Hi! I am trying to `upsert` in my table which should allow either updating an existing row, or inserting a new row if it doesn't have an ID yet. This is how I'm doing it:

```js
const { error } = await supabase
      .from('my_table')
      .upsert([{ id: 3, name: 'Abc' }, { name: 'Def' }], { onConflict: 'name' });
```

However, I get this error when I try to combine an update and an insert. It works fine if it's just all updates or if it's just all inserts.
```
Error: null value in column "id" of relation "my_table" violates not-null constraint
```

I also have the ID as an identity as shown in this:
<img width="734" alt="image" src="https://github.com/supabase/supabase/assets/13959651/17519a8d-2099-4f17-97a7-aee36433a6f8">

Would appreciate your help, thanks!

**Comment by steve-chavez** [comment, marked as answer]
Hello,

Can you try with `defaultToNull: false`?

```js
const { error } = await supabase
      .from('my_table')
      .upsert([{ id: 3, name: 'Abc' }, { name: 'Def' }], { onConflict: 'name', defaultToNull: false});
```

This is a [recent feature](https://supabase.com/blog/postgrest-11-1-release#bulk-insert-json-with-default-values) so you might have to upgrade your instance by pausing/unpausing.

**Reply by jayehernandez** [reply]
This is exactly what I needed, thank you!

**Comment by kno-raziel** [comment]
I spent time debugging a similar issue because the documentation is not clear about this option. At least not about what is the default value of the option.

 
![image](https://github.com/supabase/supabase/assets/8211946/f92eadb0-fd4b-4b4e-b44d-660994c04136)

**Reply by steve-chavez** [reply]
Hm, perhaps we should make `defaultToNull: false` the default for the next major version.

cc @soedirgo

**Reply by soedirgo** [reply]
Sounds fine DX-wise, but iirc there's a perf hit when you use it?

**Reply by soedirgo** [reply]
(Though I'd still push for it regardless)

**Comment by Cali93** [comment]
Hey, I'm facing the same issue even if I use `defaultToNull: false`.

Client version: `"@supabase/supabase-js": "^2.38.0"`

Here's my table definition
```sql
CREATE TABLE my_table (
    id UUID UNIQUE PRIMARY KEY DEFAULT uuid_generate_v4(),
    name VARCHAR(255) NOT NULL,
    description VARCHAR(3000) NOT NULL
);
```
But after running this in the Supabase editor, when I inspect the definition on the Supabase UI I see that the the table definition has a NOT NULL constraint
```sql
id uuid not null default uuid_generate_v4 (),
```
And here's the response I get if I don't provide the id on my upsert.

```json
{
  "code":"23502", 
  "details":"Failing row contains (null, Not working, Annoying).",
  "hint":null,
  "message":"null value in column \"id\" of relation \"my_table\" violates not-null constraint"
}
```

**Reply by GerritPlehn** [reply]
Make sure your Project is running a PostgREST version >= 11.1, that was the issue for me. You can check it on the "Infrastructure" tab in your project settings.
That version introduced Bulk Inserts with default values, see [release notes](https://supabase.com/blog/postgrest-11-1-release#bulk-insert-json-with-default-values)

**Reply by nicetomytyuk** [reply]
@GerritPlehn I'm using PostgREST v12.0.1 and still having the same issue.

**Reply by GerritPlehn** [reply]
@nicetomytyuk I ended up sending 2 requests, one with the records that do contain `null` and one with those who don't.
Not elegant by any means but I wasn't able to spend more time debugging this.

**Reply by nicetomytyuk** [reply]
@GerritPlehn yeah I've just finished with the same solution, something like:

![image](https://github.com/supabase/supabase/assets/30237552/1dfdfdbe-6f32-43ab-bd04-f0fa1bc1c869)
